### PR TITLE
Stop raising ERR_R_MALLOC_FAILURE in most places

### PR DESCRIPTION
**Pull request by levitte**
Since `OPENSSL_malloc()` and friends report `ERR_R_MALLOC_FAILURE`, and at least handle the file name and line number they are called from, there's no need to report `ERR_R_MALLOC_FAILURE` where they are called directly, or when `SSLfatal()` and `RLAYERfatal()` is used, the reason `ERR_R_MALLOC_FAILURE` is changed to `ERR_R_CRYPTO_LIB`.

There were a number of places where `ERR_R_MALLOC_FAILURE` was reported even though it was a function from a different sub-system that was called.  Those places are changed to report `ERR_R_{lib}_LIB`, where `{lib}` is the name of that sub-system.
Some of them are tricky to get right, as we have a lot of functions that belong in the ASN1 sub-system, and all the `sk_` calls or from the CRYPTO sub-system.

Some extra adaptation was necessary where there were custom `OPENSSL_malloc()` wrappers, and some bugs are fixed alongside these changes.
Most notably, `CRYPTO_secure_malloc()` has been changed to also raise `ERR_R_MALLOC_FAILURE` when appropriate.

This is an alternative to #19072

### CODE REVIEW

**Comment by levitte** [comment]
This is in draft for a bit, I need to re-read what I've done after a pause.
Feel free to look, comment, or even review.

**Comment by DDvO** [comment]
> This is an alternative to #19072

This is clearly more advanced than #19072.
Yet note that that PR also has commit 6a6e27b682de0aec606a61382e3822568f2c1676 which replaces all remnant `ECerr()` and `EVPerr()` calls in `crypto/` by `ERR_raise()`. This is not (yet) covered here.

**Comment by DDvO** [comment]
> That could actually be made into a completely separate fix commit. It's quite independent from all the rest

I just did this in #19302, and closed #19072 in favor of the two new PRs.

File: crypto/x509/x509_v3.c [hunk 1]
```diff
@@ -118,15 +120,13 @@ STACK_OF(X509_EXTENSION) *X509v3_add_ext(STACK_OF(X509_EXTENSION) **x,
         loc = n;
 
     if ((new_ex = X509_EXTENSION_dup(ex)) == NULL)
-        goto err2;
+        goto err;
     if (!sk_X509_EXTENSION_insert(sk, new_ex, loc))
         goto err;
```
Comment (t8m) [inline crypto/x509/x509_v3.c:125]: These look like they should have an ERR_R_X509_LIB raised.

File: ssl/ssl_sess.c [hunk 2]
```diff
@@ -198,27 +202,27 @@ SSL_SESSION *ssl_session_dup(const SSL_SESSION *src, int ticket)
     if (src->psk_identity_hint) {
         dest->psk_identity_hint = OPENSSL_strdup(src->psk_identity_hint);
         if (dest->psk_identity_hint == NULL) {
+            ERR_raise(ERR_LIB_SSL, ERR_R_X509_LIB);
```
Comment (t8m) [inline ssl/ssl_sess.c:205]: This looks wrong

**Review by t8m — CHANGES_REQUESTED** [review]
I've found just two mistakes in this gargantuan PR. How did you do it, even with some scripting, I cannot imagine. Great work!

**Comment by levitte** [comment]
> How did you do it, even with some scripting, I cannot imagine.

Grep, and eyeballing

**Comment by levitte** [comment]
I'm re-reading my changes, and clearly got blurry-eyed in some sections.  Corrections coming up!

**Comment by beldmit** [comment]
@levitte it looks impossible to review all the affected files. 
I'd suggest splitting them at least on commit per-directory basis if you want to fixup it.

File: ssl/ssl_sess.c (hunk 2 above)
Comment (levitte) [reply ssl/ssl_sess.c:205]: Yup, it was

File: crypto/x509/x509_v3.c (hunk 1 above)
Comment (levitte) [reply crypto/x509/x509_v3.c]: Yup, except `ERR_R_CRYPTO_LIB`, because that's the "sub-system" that all `sk_` functions belong in.

**Comment by levitte** [comment]
> @levitte it looks impossible to review all the affected files. I'd suggest splitting them at least on commit per-directory basis if you want to fixup it.

...  I think that'll be something for tomorrow, although I'm hoping that someone has the stamina :wink:

File: crypto/pem/pvkfmt.c [hunk 3]
```diff
@@ -88,19 +88,17 @@ static EVP_PKEY *evp_pkey_new0_key(void *key, int evp_type)
     if ((pkey = EVP_PKEY_new()) != NULL) {
         switch (evp_type) {
         case EVP_PKEY_RSA:
-            if (EVP_PKEY_set1_RSA(pkey, key)) {
-                ERR_raise(ERR_LIB_PEM, ERR_R_RSA_LIB);
+            if (EVP_PKEY_set1_RSA(pkey, key))
                 break;
-            }
+            ERR_raise(ERR_LIB_PEM, ERR_R_RSA_LIB);
```
Comment (t8m) [inline crypto/pem/pvkfmt.c:93]: Should it rather be ERR_R_EVP_LIB?

File: crypto/pem/pvkfmt.c [hunk 4]
```diff
@@ -88,19 +88,17 @@ static EVP_PKEY *evp_pkey_new0_key(void *key, int evp_type)
...
             pkey = NULL;
             break;
 #ifndef OPENSSL_NO_DSA
         case EVP_PKEY_DSA:
-            if (EVP_PKEY_set1_DSA(pkey, key)) {
-                ERR_raise(ERR_LIB_PEM, ERR_R_DSA_LIB);
+            if (EVP_PKEY_set1_DSA(pkey, key))
                 break;
-            }
+            ERR_raise(ERR_LIB_PEM, ERR_R_DSA_LIB);
```
Comment (t8m) [inline crypto/pem/pvkfmt.c:101]: And same here.

**Review by t8m — CHANGES_REQUESTED** [review]

File: crypto/pem/pvkfmt.c (hunk 3 above)
Comment (levitte) [reply crypto/pem/pvkfmt.c:93]: Ah, yeah, correct!

File: crypto/pem/pvkfmt.c (hunk 4 above)
Comment (levitte) [reply crypto/pem/pvkfmt.c:101]: Yup

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Should the draft status be removed? IMO this is good to go.

**Review by hlandau — APPROVED** [review]

**Comment by levitte** [comment]
Merged

9167a47f78159b0578bc032401ab1d66e14eecdb Adapt CRYPTO_secure_malloc() like CRYPTO_malloc()
e077455e9e57ed4ee4676996b4a9aa11df6327a6 Stop raising ERR_R_MALLOC_FAILURE in most places
79c8dcf3985a7b75eac8e53eb8652728af6c5d3d Add {lib}_R_{lib}_LIB, for our engines and other "external" modules

**Comment by wan7teen-ping** [comment]
Hi I will help you with everything you need but you have to subscribe to
the youtube channel and I will fulfill all your requests

https://youtu.be/aIaqr4eq18g

Pada Kha, 29 Sep 2022, 8:18 PM Richard Levitte ***@***.***>
menulis:

**Comment by learning310** [comment]
@levitte, I want to ask about a problem with EVP_xxx_CTX_new() in error handling and why EVP_KDF_CTX_new() is different from EVP_MAC_CTX_new(). In my view, the error handling of EVP_MAC_CTX_new() is much neater when the function is not compromised.

**Comment by bubble932** [comment]
@levitte I am curious about the differences in error handling between EVP_KDF_CTX_new() and EVP_MAC_CTX_new(). When OPENSSL_zalloc() fails, EVP_KDF_CTX_new() will raises ERR_R_EVP_LIB, whereas EVP_MAC_CTX_new() does not. Instead, EVP_MAC_CTX_new() only raises ERR_R_EVP_LIB when newctx() fails. This makes EVP_MAC_CTX_new() seem more reasonable in its error handling. Are there any other concerns I haven't noticed? Please enlighten me on this. Thanks.

**Pull request closed without merging**